### PR TITLE
Print 'Infinite' for infinite durations in topic endpoint info

### PR DESCRIPTION
**Pull request by emersonknapp**
Sample output for `ros2 topic pub /chatter std_msgs/String "data: hello"`:

```
$ ros2 topic info -v /chatter
Type: std_msgs/msg/String

Publisher count: 1

Node name: _ros2cli_30830
Node namespace: /
Topic type: std_msgs/msg/String
Endpoint type: PUBLISHER
GID: 2d.26.10.01.fb.93.9b.48.9d.7e.d3.24.00.00.08.03.00.00.00.00.00.00.00.00
QoS profile:
  Reliability: RELIABLE
  Durability: VOLATILE
  Lifespan: Infinite
  Deadline: Infinite
  Liveliness: AUTOMATIC
  Liveliness lease duration: Infinite

Subscription count: 0

```

Connected to https://github.com/ros2/ros2cli/pull/616

### CODE REVIEW

**Comment by emersonknapp** [comment]
@sloretz @ivanpauno as maintainers - requesting a review when you can

File: rclpy/rclpy/topic_endpoint_info.py
Comment (ivanpauno) [inline rclpy/rclpy/topic_endpoint_info.py:170]: feel free to ignore, but it would be much nicer to use f-strings

File: rclpy/rclpy/topic_endpoint_info.py
Comment (ivanpauno) [inline rclpy/rclpy/topic_endpoint_info.py:173]: same applies here

File: rclpy/src/rclpy/_rclpy_pybind11.cpp
Comment (ivanpauno) [inline rclpy/src/rclpy/_rclpy_pybind11.cpp:72]: can this be just named `DURATION_INFINITE`?

```suggestion
  m.attr("DURATION_INFINITE") = py::int_(rmw_time_total_nsec(RMW_DURATION_INFINITE));
```

The prefixes are ok for C macros which polute the global namespace but aren't needed here (`_rclpy.DURATION_INFINITE` is clear enough).

**Review by ivanpauno — APPROVED** [review]

File: rclpy/rclpy/topic_endpoint_info.py
Comment (emersonknapp) [reply rclpy/rclpy/topic_endpoint_info.py:170]: I was feeling that if I update one to f-strings, I should update them all at the same time, but wasn't sure if that was an overreach on the scope of this PR - what do you think?

File: rclpy/src/rclpy/_rclpy_pybind11.cpp
Comment (emersonknapp) [reply rclpy/src/rclpy/_rclpy_pybind11.cpp:72]: I agree with this assessment. However, the way that the current pybind code works, anything that doesn't start with `rclpy_` is ignored in the module merge (https://github.com/aws-ros-dev/rclpy/blob/emersonknapp/topic-info-duration-infinite/rclpy/rclpy/impl/implementation_singleton.py#L46), so I had to add an `rclpy_` prefix.

Comment (ivanpauno) [reply rclpy/src/rclpy/_rclpy_pybind11.cpp:72]: @sloretz what do you think?

File: rclpy/rclpy/topic_endpoint_info.py
Comment (ivanpauno) [reply rclpy/rclpy/topic_endpoint_info.py:170]: > what do you think?

sounds fine to me to do it here

File: rclpy/src/rclpy/_rclpy_pybind11.cpp
Comment (emersonknapp) [reply rclpy/src/rclpy/_rclpy_pybind11.cpp:72]: @sloretz friendly ping! should we consider changing the pybind name semantics to expose constants, or keep this as-is?

Comment (emersonknapp) [reply rclpy/src/rclpy/_rclpy_pybind11.cpp]: Based on the latest code, I was able to remove the prefix from this constant

**Comment by emersonknapp** [comment]
@ivanpauno requesting re-review now that I was able to address all the comments, given the latest state of the pybind code

**Review by sloretz — APPROVED** [review]

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/8243fb1b94c37e0c9a07ec4c5edfb2a5/raw/bc58898af9aa8ad525c5d97279271f0b8c9f4734/ros2.repos
BUILD args: --packages-up-to rclpy
TEST args: --packages-select rclpy
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/8046
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14161/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8960/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11841/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14259/)

**Review by ivanpauno — APPROVED** [review]

**Comment by ivanpauno** [comment]
Running tests above `rclpy`, I'm afraid some `ros2cli` test might depend on how a duration is printed:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14164/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8963/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11844/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14263/)

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14178/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8975/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11859/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14278/)

**Comment by ivanpauno** [comment]
> CI with packages-above rclpy testing, including ros2/ros2cli#616

It doesn't seem to be including ros2/ros2cli#616, again including it:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14235/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9033/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11926/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14339/)

**Comment by emersonknapp** [comment]
I think the failures were because my branches were out of date with the fast development on rclpy right now

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14242/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9040/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11933/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14347/)

**Comment by ivanpauno** [comment]
^ note: rviz_ogre_vendor timed out build on Windows - should I just try it again?

:+1:

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14350/)

**Comment by emersonknapp** [comment]
Warnings from https://github.com/eclipse-cyclonedds/cyclonedds/issues/741

**Comment by ivanpauno** [comment]
Sorry, I wasn't available yesterday late (I work on UTC-3 time zone) and now we're past the API/feature freeze.

I'm not completely sure if this passes the post API/feature "safe to merge" criteria, I'm in favor of merging it as it makes `ros2cli` output more clear and I don't think this can break downstream code.

@hidmic could you do a second review?

**Review by audrow — APPROVED** [review]

**Comment by emersonknapp** [comment]
Thanks @audrow ! Does this mean we're planning to get this into Galactic?

**Comment by audrow** [comment]
I believe so. Let me just confirm with someone more senior on the team.

**Comment by hidmic** [comment]
> @hidmic could you do a second review?

Argh, I didn't see this. My apologies.

File: rclpy/rclpy/duration.py
Comment (hidmic) [inline rclpy/rclpy/duration.py:87]: @emersonknapp nit: considering `Duration` objects are immutable, to have a single `Infinite` instance would be best.

**Review by hidmic — APPROVED** [review]
LGTM, but the line between bug fix and feature is quite blurry here. Feature freeze was almost two weeks ago. If we merge this for Galactic, we'd have to make an exception. @cottsay ?

**Comment by cottsay** [comment]
> If we merge this for Galactic, we'd have to make an exception

I agree. Can we wait until we branch galactic next week to merge this into rolling? And then consider it for backport into Galactic after we release it?

File: rclpy/rclpy/duration.py
Comment (emersonknapp) [reply rclpy/rclpy/duration.py]: sure

**Comment by emersonknapp** [comment]
Whatever y'all want to do

**Comment by ivanpauno** [comment]
This can be merged now.
Running an extra round of CI, just in case:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14577/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9332/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12239/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14704/)

**Comment by DLu** [comment]
Note: I came here to fix this, not realizing the problem was that I was running `foxy`. Thanks! Still, might be worth documenting: 
https://github.com/ros2/ros2_documentation/pull/4157

**Pull request merged**
